pagein.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PAGEIN 1 "12 January 2024"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
pagein \- a tool to force swapped out pages back into memory
.br

.SH SYNOPSIS
.B pagein
.RI [ \-a ]
.RI [ \-h ]
.RI [ \-p\ pid ]
.RI [ \-v ]
.br

.SH DESCRIPTION
Pagein is a tool that forces pages that are in swap to be paged in back
to memory.  The main usecase for pagein is to exercise the VM and swap
subsystems for testing purposes.

.SH OPTIONS
pagein options are as follow:
.TP
.B \-a
try to page in all processes. This needs to be run with root privilege to
allow pagein to access all the processes. Cannot be used with the \-p
option.
.TP
.B \-h
show help information.
.TP
.B \-p pid
page in process with process ID pid. Cannot be used with the \-a option.
Processes not owned by the user need to be paged in by running pagein
with root privilege.
.TP
.B \-v
verbose mode, shows progress and paging in statistics, useful with the
\-a option.
.SH EXAMPLES
.LP
Page in all processes and show progress:
.RS 8
sudo pagein \-a \-v
.RE
.LP
Page in the Xorg process:
.RS 8
sudo pagein \-p $(pidof Xorg)
.RE
.SH SEE ALSO
.BR free (1)
.SH AUTHOR
pagein was written by Colin Ian King <colin.i.king@gmail.com>
.PP
This manual page was written by Colin Ian King
for the Ubuntu project (but may be used by others).
.SH COPYRIGHT
Copyright \(co 2016-2021 Canonical Ltd, Copyright \(co 2021-2025 Colin Ian King.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


debian/changelog
pagein (0.01.11-1) unstable; urgency=medium

  * Makefile: bump version
  * debian/control: update Standards-Version to 4.7.0
  * pagein: use off64_t for fdmem seeks, check for -ve offsets

 -- Colin Ian King <colin.i.king@gmail.com>  Wed, 17 Jul 2024 16:58:14 +0100

pagein (0.01.10-1) unstable; urgency=medium

  * Makefile: bump version
  * Makefile: use xz compression on tarball on make dist rule
  * debian/control: update standards version
  * Update copyright year
  * debian/rules: add missing hardening flags
  * Makefile: remove snap from make dist rule
  * Remove snapcraft

 -- Colin Ian King <colin.i.king@gmail.com>  Fri, 12 Jan 2024 11:13:32 +0000

pagein (0.01.09-1) unstable; urgency=medium

  * Makefile: bump version
  * debian/control: update standards version to 4.6.1
  * Add comm field information in verbose mode, print pids with %jd
  * Update copyright year

 -- Colin Ian King <colin.i.king@gmail.com>  Wed, 14 Dec 2022 13:26:15 +0000

pagein (0.01.08-1) unstable; urgency=medium

  * Debian: update changelog
  * Makefile: update copyright, bump version
  * Fix potential array overflow on array prot, make it 1 byte larger

 -- Colin Ian King <colin.i.king@gmail.com>  Wed, 20 Apr 2022 12:01:03 +0100

pagein (0.01.07-1) unstable; urgency=medium

  * Makefile: bump version
  * pagein.c: update copyright
  * Makefile: update copyright
  * Makefile: add README.md to dist rule
  * manual: update date, author email and copyright
  * debian/watch: update watch path for tarball, move to version 4
  * debian/copyright: Update copyright and git URL
  * debian/control: update author email address and project URL
  * Add README.md
  * snapcraft: update to core20

 -- Colin Ian King <colin.i.king@gmail.com>  Mon, 15 Nov 2021 19:51:23 +0000

pagein (0.01.06-1) unstable; urgency=medium

  * Makefile: bump version
  * Ensure memfree_* and pagefree_* are initialized
  * update copyright to 2021
  * snapcraft.yaml: fix versioning magic
  * snapcraft.yaml: update snapcraft versioning magic

 -- Colin King <colin.king@canonical.com>  Wed, 28 Apr 2021 20:54:15 +0100

pagein (0.01.05-1) unstable; urgency=medium

  * Makefile: bump version
  * Debian/control: update compat to 13, remove compat file,
    add Rules-Requires-Root rule

 -- Colin King <colin.king@canonical.com>  Sat, 4 Jul 2020 15:39:22 +0100

pagein (0.01.04-1) unstable; urgency=medium

  * Makefile: update version
  * Add bash command completion script
  * Update copyright to 2020
  * snapcraft: add github URL for project
  * Makefile: rename snapcraft to snap in dist rule
  * Rename snapcraft to snap

 -- Colin King <colin.king@canonical.com>  Wed, 26 Feb 2020 14:07:11 +0000

pagein (0.01.03-1) unstable; urgency=medium

  * Makefile: bump version
  * Debian: update compat to level 12
  * snapcraft: use automatic versioning

 -- Colin King <colin.king@canonical.com>  Mon, 12 Aug 2019 16:47:25 +0100

pagein (0.01.02-1) unstable; urgency=medium

  * Makefile: remove -g flag, bump version
  * Makefile: add .travis.yml to dist rule
  * Update copyright year
  * Add travis yaml file

 -- Colin King <colin.king@canonical.com>  Wed, 10 Jul 2019 10:41:23 +0100

pagein (0.01.01-1) unstable; urgency=medium

  * Makefile: bump version
  * Remove kthreads stat, add in total mappings stat
  * Add unreadable skipped pages stats
  * Expand pid field to 6 digits
  * Manual: change example to Xorg as a process to page in
  * Remove defunct command options -d, -t, -w
  * debian/copyright: use secure URI
  * debian/control: remove empty line and set debhelper version to >=11
  * debian/compat: update to 11

 -- Colin King <colin.king@canonical.com>  Sun, 17 Jun 2018 12:07:00 +0100

pagein (0.01.00-1) unstable; urgency=medium

  * Makefile: bump version
  * Remove whitespace in text
  * Put altstack on main's stack, it is small enough
  * Make ptrace halt window smaller
  * Rework, just use ptrace memory peeking to force pages in
  * Add in some more constifications
  * Minor code style changes, make some small helper funcs inline too
  * Update copyright year

 -- Colin King <colin.king@canonical.com>  Sun, 14 Jan 2018 16:21:12 +0000

pagein (0.00.06-1) unstable; urgency=medium

  * Makefile: bump version
  * Make function align_address static
  * debian/control: update Standards-Version to 4.1.2
  * Rename array path to tmppath to avoid name shadowing

 -- Colin King <colin.king@canonical.com>  Thu, 4 Jan 2018 23:59:23 +0000

pagein (0.00.05-1) unstable; urgency=medium

  * Makefile: bump version again
  * Debian/control: update standards to 4.0.0
  * Remove exit after a goto, fixes structurally dead code warning
  * Make variable sz static so longjmp cannot clobber it
  * Make file mmapping private
  * Add SEGV handler and fix length calculation
  * Seek and read correct page, reference on the non-mmapp'd addresses
  * Optimize pages touched count
  * Simplify the zero length check
  * Fix another segv issue, touch the mapped area

 -- Colin King <colin.king@canonical.com>  Wed, 21 Jun 2017 08:26:18 +0100

pagein (0.00.04-1) unstable; urgency=medium

  * Makefile: bump version
  * Fix segfault on i386 targets (Closes: #865341)
  * snapcraft: add default app and grade keys

 -- Colin King <colin.king@canonical.com>  Tue, 20 Jun 2017 19:10:23 +0100

pagein (0.00.03-1) unstable; urgency=medium

  * Initial Debian release (Closes: #856462)
  * Makefile: bump version
  * Add more explanation to why access failures can occur
  * ptrace should be performed before the open of /proc/$pid/mem
  * Debian/control: fix typo on pagein name

 -- Colin King <colin.king@canonical.com>  Sat, 4 Mar 2017 20:52:33 +0000

pagein (0.00.02-1) unstable; urgency=medium

  * Makefile: bump version
  * Minor reworking of page in mapping, clean up readability
  * Fixup some build warnings on 32 bit arches
  * Makefile: update dist rules to incorporate snapcraft files
  * Add COPYING file

 -- Colin King <colin.king@canonical.com>  Wed, 1 Mar 2017 10:08:03 +0000


pagein.c
/*
 * Copyright (C) 2016-2021 Canonical, Ltd.
 * Copyright (C) 2021-2025 Colin Ian King
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

#define _LARGEFILE64_SOURCE

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <inttypes.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/resource.h>
#include <sys/ptrace.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <limits.h>
#include <errno.h>
#include <dirent.h>
#include <ctype.h>
#include <setjmp.h>
#include <signal.h>

#define APP_NAME		"pagein"
#define PAGE_4K			(4096)

#define STACK_ALIGNMENT		(64)

#define OPT_VERBOSE		(0x00000001)
#define OPT_ALL			(0x00000002)
#define OPT_BY_PID		(0x00000004)

#define GOT_MEMFREE		(0x01)
#define GOT_SWAPFREE		(0x02)
#define GOT_ALL			(GOT_MEMFREE | GOT_SWAPFREE)

#define COMM_FIELD_SIZE		(16)

static uint16_t		opt_flags;
static sigjmp_buf 	jmp_env;

static void sigsegv_handler(int sig)
{
	static bool faulted = false;

	(void)sig;

	if (!faulted) {
		faulted = true;		/* Don't double fault */
		siglongjmp(jmp_env, 1);
	}
}

/*
 *  get_comm()
 *	get the process comm field data, ensure
 *	non-printable chars are filtered out
 */
static inline char *get_comm(const pid_t pid)
{
	char buf[PATH_MAX];
	int fd;
	ssize_t ret;
	static char comm[COMM_FIELD_SIZE + 1];
	static char unknown[] = "(unknown)";
	char *ptr;

	(void)snprintf(buf, sizeof(buf), "/proc/%jd/comm", (intmax_t)pid);
	fd = open(buf, O_RDONLY);
	if (fd < 0)
		return unknown;
	ret = read(fd, comm, COMM_FIELD_SIZE);
	(void)close(fd);
	if (ret < 0)
		return unknown;
	if (ret > COMM_FIELD_SIZE)
		ret = COMM_FIELD_SIZE;
	comm[ret] = '\0';

	for (ptr = comm; *ptr; ptr++) {
		if (*ptr == '\n') {
			*ptr = '\0';
			break;
		}
		if (!isprint((int)*ptr))
			*ptr = ' ';
	}
	return comm;
}

/*
 *  align_address
 *	align address to alignment, alignment MUST be a power of 2
 */
static inline void *align_address(const void *addr, const size_t alignment)
{
	const uintptr_t uintptr =
		((uintptr_t)addr + alignment) & ~(alignment - 1);

	return (void *)uintptr;
}

/*
 *  get_page_size()
 *	get page size
 */
static inline int32_t get_page_size(void)
{
	int32_t page_size;
#ifdef _SC_PAGESIZE
	static int32_t sz;	/* static noclobber */

	sz = sysconf(_SC_PAGESIZE);
	page_size = (sz <= 0) ? PAGE_4K : sz;
#else
	page_size = PAGE_4K;
#endif
	return page_size;
}

/*
 *  get_memstats()
 *	get some pertinent memory statistics from /proc/meminfo
 */
static int get_memstats(int64_t *const memfree, int64_t *const swapfree)
{
	FILE *fp;
	char buffer[4096];
	uint8_t got = 0;

	*memfree = 0;
	*swapfree = 0;

	fp = fopen("/proc/meminfo", "r");
	if (!fp)
		return -errno;

	while (fgets(buffer, sizeof(buffer), fp)) {
		if ((strstr(buffer, "MemFree:")) &&
		    (sscanf(buffer, "%*s %" SCNd64, memfree) == 1)) {
			got |= GOT_MEMFREE;
			continue;
		}
		if ((strstr(buffer, "SwapFree:")) &&
		    (sscanf(buffer, "%*s %" SCNd64, swapfree) == 1)) {
			got |= GOT_SWAPFREE;
		}
		if ((got & GOT_ALL) == GOT_ALL)
			break;
	}

	(void)fclose(fp);
	return 0;
}

/*
 *  show_help()
 *	show command help info
 */
static void show_help(void)
{
	(void)printf(APP_NAME ":\n");
	(void)printf("Usage: " APP_NAME " [OPTION [ARG]]\n");
	(void)printf("-a\tpage in pages in all processes\n");
	(void)printf("-h\tshow help\n");
	(void)printf("-p pid\tpull in pages on specified process\n");
	(void)printf("-v\tverbose mode\n");
	(void)printf("Note: to page in all processes, run with root privilege\n");
}

/*
 *  pagein_proc()
 *	try to force page in pages for a specific process
 */
static int pagein_proc(
	const int32_t page_size,
	const pid_t pid,
	int32_t *const procs,
	int64_t *const total_pages_touched,
	int64_t *const total_pages_skipped,
	int64_t *const total_mappings)
{
	char path[PATH_MAX];
	char buffer[4096];
	int fdmem, rc = 0, ret;
	FILE *fpmap;
	size_t pages = 0, pages_touched = 0;
	bool has_maps = false;

	if (pid == getpid())
		return 0;

	(void)snprintf(path, sizeof(path), "/proc/%d/mem", pid);
	fdmem = open(path, O_RDONLY);
	if (fdmem < 0) {
		rc = -errno;
		goto err_ret;
	}

	(void)snprintf(path, sizeof(path), "/proc/%d/maps", pid);
	fpmap = fopen(path, "r");
	if (!fpmap) {
		rc = -errno;
		goto err_close_fdmem;
	}

	ret = ptrace(PTRACE_ATTACH, pid, NULL, NULL);
	if (ret < 0) {
		rc = -errno;
		goto err_close_all;
	}

	(void)waitpid(pid, NULL, 0);

	/*
	 * Look for field 0060b000-0060c000 r--p 0000b000 08:01 1901726
	 */
	while (fgets(buffer, sizeof(buffer), fpmap)) {
		uint64_t begin, end, len;
		uintptr_t off;
		char tmppath[1024];
		char prot[6];

		if (sscanf(buffer, "%" SCNx64 "-%" SCNx64
			   " %5s %*x %*x:%*x %*d %1023s", &begin, &end, prot, tmppath) != 4)
			continue;
		(*total_mappings)++;
		len = end - begin;
		if (prot[0] != 'r' && prot[3] != 'p') {
			*total_pages_skipped += (len / page_size);
			continue;
		}

		if ((begin >= end) || (len == 0))
			continue;

		has_maps = true;

		for (off = begin; off < end; off += page_size, pages++) {
			unsigned long data;
			off64_t off64;

			(void)ptrace(PTRACE_PEEKDATA, pid, (void *)off, &data);

			off64 = (off64_t)off;
			if (off64 >= 0)  {
				off64_t pos64;

				pos64 = lseek64(fdmem, off64, SEEK_SET);
				if (pos64 == off64) {
					size_t sz;

					sz = read(fdmem, &data, sizeof data);
					(void)sz;
				}
			}
			pages_touched++;
		}
	}
	(void)ptrace(PTRACE_DETACH, pid, NULL, NULL);

	/*
	 *  Kernel threads don't have maps, so don't
	 *  count these in the stats
	 */
	if (has_maps) {
		(*procs)++;
		*total_pages_touched += pages_touched;
	}

	if (opt_flags & OPT_VERBOSE) {
		(void)printf("PID:%8jd %-*.*s %12zu pages, %12" PRId64 " pages touched\r",
			(intmax_t)pid, COMM_FIELD_SIZE, COMM_FIELD_SIZE, get_comm(pid),
			pages, *total_pages_touched);
		(void)fflush(stdout);
	}

err_close_all:
	(void)fclose(fpmap);
err_close_fdmem:
	(void)close(fdmem);
err_ret:
	return rc;
}

/*
 *  pagein_all_procs()
 *	attempt to page in all processes
 */
static inline int pagein_all_procs(
	const int32_t page_size,
	int32_t *const procs,
	int32_t *const total_procs,
	int64_t *const total_pages_touched,
	int64_t *const total_pages_skipped,
	int64_t *const total_mappings)
{
	DIR *dp;
	struct dirent *d;

	dp = opendir("/proc");
	if (!dp)
		return -1;

	while ((d = readdir(dp)) != NULL) {
		pid_t pid;

		if (isdigit(d->d_name[0]) &&
                    sscanf(d->d_name, "%d", &pid) == 1) {
			*total_procs += 1;
			pagein_proc(page_size, pid, procs,
				total_pages_touched, total_pages_skipped,
				total_mappings);
		}
	}

	(void)closedir(dp);

	return 0;
}

int main(int argc, char **argv)
{
	int64_t memfree_begin = 0, memfree_end = 0;
	int64_t swapfree_begin = 0, swapfree_end = 0;
	int64_t delta;
	int64_t total_pages_touched = 0ULL;
	int64_t total_pages_skipped = 0ULL;
	int64_t total_mappings = 0ULL;
	static int32_t total_procs = 0;		/* static noclobber */
	int32_t procs = 0;
	const int32_t page_size = get_page_size();
	const int32_t scale = page_size / 1024;
	static pid_t pid = -1;			/* static noclobber */
	struct rusage usage;
	struct sigaction action;
	stack_t ss;
	uint8_t	stack[SIGSTKSZ + STACK_ALIGNMENT];

	for (;;) {
		int c = getopt(argc, argv, "ahp:v");
		if (c == -1)
			break;

		switch (c) {
		case 'a':
			opt_flags |= OPT_ALL;
			break;
		case 'h':
			show_help();
			exit(EXIT_SUCCESS);
			break;
		case 'p':
			opt_flags |= OPT_BY_PID;
			pid = atoi(optarg);
			if (pid < 1) {
				(void)fprintf(stderr, "bad pid: %d\n", pid);
				exit(EXIT_FAILURE);
			}
			break;
		case 'v':
			opt_flags |= OPT_VERBOSE;
			break;
		default:
			show_help();
			exit(EXIT_FAILURE);
		}
	}

	if ((opt_flags & (OPT_ALL | OPT_BY_PID)) ==
		(OPT_ALL | OPT_BY_PID)) {
		(void)fprintf(stderr, "must not use -a and -p options together\n");
		exit(EXIT_FAILURE);
	}
	if ((opt_flags & (OPT_ALL | OPT_BY_PID)) == 0) {
		(void)fprintf(stderr, "must use one of -a or -p, use -h for more details\n");
		exit(EXIT_FAILURE);
	}

	if (sigsetjmp(jmp_env, 1) == 1) {
		(void)printf("Aborted early, hit a page fault\n");
		goto finish;
	}

	ss.ss_sp = align_address(&stack, STACK_ALIGNMENT);
	ss.ss_size = SIGSTKSZ;
	ss.ss_flags = 0;
	if (sigaltstack(&ss, NULL) < 0) {
		(void)fprintf(stderr, "cannot set sigaltstack stack\n");
		exit(EXIT_FAILURE);
	}
	(void)memset(&action, 0, sizeof(action));
	(void)sigemptyset(&action.sa_mask);
	action.sa_handler = sigsegv_handler;
	action.sa_flags = SA_ONSTACK;

	if (sigaction(SIGSEGV, &action, NULL) < 0) {
		(void)fprintf(stderr, "cannot set signal handler\n");
		exit(EXIT_FAILURE);
	}
		
	get_memstats(&memfree_begin, &swapfree_begin);
	if (opt_flags & OPT_ALL)
		pagein_all_procs(page_size, &procs,
				 &total_procs, &total_pages_touched,
				 &total_pages_skipped, &total_mappings);

	if (opt_flags & OPT_BY_PID) {
		int ret;

		ret = pagein_proc(page_size, pid, &procs,
			&total_pages_touched, &total_pages_skipped,
			&total_mappings);
		if (ret < 0) {
			(void)fprintf(stderr, "cannot page in PID %d errno = %d (%s)\n",
				pid, -ret, strerror(-ret));
			(void)fprintf(stderr, "  Note: this is normally because of ptrace PTRACE_MODE_ATTACH_FSCREDS access\n");
			(void)fprintf(stderr, "  mode failure of pagein. pagein needs to be run with the CAP_SYS_PTRACE\n");
			(void)fprintf(stderr, "  capability (for example, run pagein as root).\n");
			exit(EXIT_FAILURE);
		}
	}

	get_memstats(&memfree_end, &swapfree_end);
	if (opt_flags & OPT_VERBOSE)
		(void)printf("%-60.60s\r", "");

finish:
	if (opt_flags & OPT_ALL) {
		(void)printf("Processes scanned:     %" PRIu32 "\n", total_procs);
		(void)printf("Processes touched:     %" PRIu32 "\n", procs);
	}
	(void)printf("Page mappings visited: %" PRIu64 "\n", total_mappings);
	(void)printf("Pages touched:         %" PRIu64 "\n", total_pages_touched);
	(void)printf("Pages skipped:         %" PRIu64 " (unreadable)\n", total_pages_skipped);
	delta = memfree_begin - memfree_end;
	(void)printf("Free memory decrease:  %" PRId64 "K (%" PRId64 " pages)\n",
		delta, delta / scale);
	delta = swapfree_begin - swapfree_end;
	(void)printf("Swap memory decrease:  %" PRId64 "K (%" PRId64 " pages)\n",
		delta, delta / scale);
	if (getrusage(RUSAGE_SELF, &usage) == 0) {
		(void)printf("Page faults major:     %lu\n", usage.ru_majflt);
		(void)printf("Page faults minor:     %lu\n", usage.ru_minflt);
		(void)printf("Swaps:                 %lu\n", usage.ru_nswap);
	}

	return EXIT_SUCCESS;
}
